File: heap_sort.cpp
#include <stdio.h>
#include <iostream>
void swap(int tree[], int m, int n)
{
   int temp = tree[m];
   tree[m] = tree[n];
   tree[n] = temp;
}
void heapify(int tree[], int n, int i)
{  
   if(i>=n)
   {
     return;
   }	   
   int left= 2 * i + 1;
   int right= 2 * i + 2;
   int max = i;
   if(left<n && tree[left]>tree[max])
   {
      max = left;
   }
   if (right<n && tree[right]>tree[max])
   {
      max = right;
   }
   if(max!=i)
   {
      swap(tree, max, i);
	  heapify(tree,n, max);
   }
}

void build_heap(int tree[], int n)
{
   int last_node = n-1;
   int parent = (last_node - 1)/2;
   for(int i=parent;i>=0;i--)
   {
	   heapify(tree,n,i);
   }

}

void heap_sort(int tree[], int n)
{
  for(int i =n;i>0;i--)
  {
	  build_heap(tree,i);
      swap(tree,0,i-1);
  }
}

int main()
{
  int tree[]={3,9,2,4,8};
  heap_sort(tree,5);
  //build_heap(tree,5);
  //heapify(tree, 5, 0);
  for(int i=0;i<5;i++)
	 std::cout<<tree[i]<<std::endl;
  return 0;  
}

File: heap_sort.py
def heapify(arr, size, i):
    if i>=size:
        return
    max = i;
    left = 2*i + 1
    right = 2*i +2
    if(left<size and arr[left]>=arr[max]):
         max = left
    if (right<size and arr[right]>=arr[max]):
         max = right
    if max!=i:
         arr[i], arr[max] = arr[max], arr[i]
         heapify(arr, size, max)
def build_heap(arr,size):
    last_node = size -1
    parent = (last_node-1)//2
    for i in range(parent,-1,-1):
        heapify(arr,size, i)

def heap_sort(arr, size):
    for i in range(size,1,-1):
        build_heap(arr, i)
        arr[i-1], arr[0] = arr[0], arr[i-1]
        print(arr)




a = [9,11,23,4,5,89,100]

#heapify(a,7,0)
#print(a)
#build_heap(a,7)
#print(a)
heap_sort(a,7)
print(a)






